piper/ledbutton.py

```diff
@@ -0,0 +1,181 @@
+                             "LED_MODE_BREATHING are supported")
+        self.adjustment_brightness.set_value(self._led.brightness)
+        self.adjustment_effect_rate.set_value(self._led.effect_rate)
+
+    @GObject.Property
+    def rgba(self):
+        """The selected color, as Gdk.RGBA."""
+        return self.colorbutton.get_rgba()
+
+    @GObject.Property
+    def brightness(self):
+        """The selected brightness value, as int."""
+        return self.adjustment_brightness.get_value()
+
+    @GObject.Property
+    def effect_rate(self):
+        """The selected effect rate, as int."""
+        return self.adjustment_effect_rate.get_value()
+
+
+@GtkTemplate(ui="/org/freedesktop/Piper/ui/LedButton.ui")
+class LedButton(Gtk.ButtonBox):
+    """A Gtk.ButtonBox subclass to implement the toggle buttons that set an
+    LED's mode."""
+
+    __gtype_name__ = "LedButton"
+
+    button_solid = GtkTemplate.Child()
+    button_cycle = GtkTemplate.Child()
+    button_breathing = GtkTemplate.Child()
+    button_off = GtkTemplate.Child()
+
+    def __init__(self, ratbagd_led, *args, **kwargs):
+        """Instantiates a new LedButton.
+
+        @param ratbagd_led The LED to configure, as ratbagd.RatbagdLed
+        """
+        Gtk.ButtonBox.__init__(self, *args, **kwargs)
+        self.init_template()
+        self._led = ratbagd_led
+        self._init_ui()
+
+    def _init_ui(self):
+        # Activate the button that maps to the LED's current mode
+        mode = self._led.mode
+        if mode == RatbagdLed.LED_MODE_OFF:
+            self.button_off.set_active(True)
+        elif mode == RatbagdLed.LED_MODE_ON:
+            self.button_solid.set_active(True)
+        elif mode == RatbagdLed.LED_MODE_CYCLE:
+            self.button_cycle.set_active(True)
+        elif mode == RatbagdLed.LED_MODE_BREATHING:
+            self.button_breathing.set_active(True)
+
+        # Only connect to the signals now so we don't map a dialog when the main
+        # window isn't up yet.
+        self.button_off.connect("clicked", self._on_button_clicked, RatbagdLed.LED_MODE_OFF)
+        self.button_solid.connect("clicked", self._on_button_clicked, RatbagdLed.LED_MODE_ON)
+        self.button_cycle.connect("clicked", self._on_button_clicked, RatbagdLed.LED_MODE_CYCLE)
+        self.button_breathing.connect("clicked", self._on_button_clicked, RatbagdLed.LED_MODE_BREATHING)
+
+    def _on_button_clicked(self, button, mode):
+        # Responds to a click on the button. The button internally always
+        # toggles on a click, so we cannot toggle only when the user applies
+        # changes in the dialog and abort when the user cancels the dialog. As
+        # such, we always change the mode; this is probably fine since the user
+        # has to commit the changes to the device anyway, but it's not ideal.
+        if not button.get_active():
+            # Prevent the unactivated togglebutton's dialog from showing.
```

Comment:
I mean that if button A is toggled, and now you toggle button B, then both A's and B's dialogs appear. This is because both A and B emit the `toggled` signal; A because it was deactivated and B because it was activated. This check is there so that a dialog only appear if a button is activated from the toggle, i.e. so that only button B's dialog appears.